subsys/usb/device_next/class/usbd_midi.c

```diff
@@ -0,0 +1,786 @@
+	struct usb_desc_header const **hs_descs;
+};
+
+/* Device driver data */
+struct usbd_midi_data {
+	struct usbd_class_data *class_data;
+	struct k_work rx_work;
+	struct k_work tx_work;
+	uint8_t tx_queue_buf[MIDI_QUEUE_SIZE];
+	struct ring_buf tx_queue;
+	uint8_t altsetting;
+	struct usbd_midi_ops ops;
+};
+
+static void usbd_midi2_recv(const struct device *dev, struct net_buf *const buf)
+{
+	struct usbd_midi_data *data = dev->data;
+	struct midi_ump ump;
+
+	LOG_HEXDUMP_DBG(buf->data, buf->len, "MIDI2 - Rx DATA");
+	while (buf->len >= 4) {
+		ump.data[0] = net_buf_pull_le32(buf);
+		for (size_t i = 1; i < UMP_NUM_WORDS(ump); i++) {
+			if (buf->len < 4) {
+				LOG_ERR("Incomplete UMP");
+				return;
+			}
+			ump.data[i] = net_buf_pull_le32(buf);
+		}
+
+		if (data->ops.rx_packet_cb) {
+			data->ops.rx_packet_cb(dev, ump);
+		}
+	}
+
+	if (buf->len) {
+		LOG_HEXDUMP_WRN(buf->data, buf->len, "Trailing data in Rx buffer");
+	}
+}
+
+static int usbd_midi_class_request(struct usbd_class_data *const class_data,
+				   struct net_buf *const buf, const int err)
+{
+	struct usbd_context *uds_ctx = usbd_class_get_ctx(class_data);
+	const struct device *dev = usbd_class_get_private(class_data);
+	struct usbd_midi_data *data = dev->data;
+	struct udc_buf_info *info = udc_get_buf_info(buf);
+
+	LOG_DBG("USB-MIDI request for %s ep=%d len=%d err=%d", dev->name, info->ep, buf->len, err);
+
+	if (err && err != -ECONNABORTED) {
+		LOG_ERR("Transfer error %d", err);
+	}
+	if (USB_EP_DIR_IS_OUT(info->ep)) {
+		usbd_midi2_recv(dev, buf);
+		k_work_submit(&data->rx_work);
+	} else {
+		LOG_HEXDUMP_DBG(buf->data, buf->len, "Tx DATA complete");
+		if (ring_buf_size_get(&data->tx_queue)) {
+			k_work_submit(&data->tx_work);
+		}
+	}
+
+	return usbd_ep_buf_free(uds_ctx, buf);
+}
+
+static void usbd_midi_class_update(struct usbd_class_data *const class_data,
+				   const uint8_t iface, const uint8_t alternate)
+{
+	const struct device *dev = usbd_class_get_private(class_data);
+	enum usbd_midi_status status = USBD_MIDI_DISABLED;
+	struct usbd_midi_data *data = dev->data;
+
+	LOG_DBG("update for %s: if=%u, alt=%u", dev->name, iface, alternate);
+
+	switch (alternate) {
+	case MIDI1_ALTERNATE:
+		data->altsetting = MIDI1_ALTERNATE;
+		LOG_WRN("%s set USB-MIDI1.0 altsetting (not implemented !)", dev->name);
+		break;
+	case MIDI2_ALTERNATE:
+		data->altsetting = MIDI2_ALTERNATE;
+		status = USBD_MIDI_2_ENABLED;
+		LOG_INF("%s set USB-MIDI2.0 altsetting", dev->name);
+		break;
+	}
+
+	if (data->ops.status_change_cb) {
+		data->ops.status_change_cb(dev, status);
+	}
+}
+
+static void usbd_midi_class_enable(struct usbd_class_data *const class_data)
+{
+	const struct device *dev = usbd_class_get_private(class_data);
+	struct usbd_midi_data *data = dev->data;
+
+	if (data->altsetting == MIDI2_ALTERNATE && data->ops.status_change_cb) {
+		data->ops.status_change_cb(dev, USBD_MIDI_2_ENABLED);
+	}
+
+	LOG_DBG("USB-MIDI enable for %s", dev->name);
```

Comment:
Prefixing logging message with USB-MIDI does not make any sense, they are already prefixed with `usbd_midi2:`

LOG_DBG("Enable %s", dev->name);